patroni/validator.py

```diff
@@ -0,0 +1,228 @@
+                                             config["postgresql"]["listen"], e)
+                connect_address = config["postgresql"].get("connect_address",
+                                                           config["postgresql"]["listen"].split(",")[0])
+                if not isinstance(connect_address, str):
+                    logger.warning("postgresql.connect_address is not a string")
+                else:
+                    connect_host, _ = split_host_port(connect_address, 5432)
+                    if connect_host in ["localhost", "127.0.0.1", "0.0.0.0", "::1", "*"]:
+                        logger.warning("postgresql.connect_address has wrong host part(%s)", connect_host)
+
+            if "parameters" not in config["postgresql"]:
+                logger.warning("postgresql.parameters is not defined")
+            elif not isinstance(config["postgresql"]["parameters"], dict):
+                logger.warning("postgresql.parameters is not a dictionary")
+            else:
+                if "unix_socket_directories" not in config["postgresql"]["parameters"]:
+                    logger.warning("postgresql.parameters.unix_socket.directories is not defined")
+                elif not isinstance(config["postgresql"]["parameters"]["unix_socket_directories"], str):
+                    logger.warning("postgresql.parameters.unix_socket_directories is not a string")
+
+    if "restapi" not in config:
+        logger.warning("restapi section is missing")
+    elif not isinstance(config["restapi"], dict):
+        logger.warning("restapi is not a dictionary")
+    elif "listen" not in config["restapi"]:
+        logger.warning("restapi.listen is not defined")
+    elif not isinstance(config["restapi"]["listen"], str):
+        logger.warning("restapi.listen is not a string")
+    else:
+        try:
+            host, port = split_host_port(config["restapi"]["listen"], None)
+        except (ValueError, TypeError):
+            logger.error("restapi.listen contains a wrong value: '%s'", config["restapi"]["listen"])
+        else:
+            with socket.socket(socket.AF_INET, socket.SOCK_STREAM) as s:
+                try:
+                    if s.connect_ex((host, port)) == 0:
+                        logger.warning("Port %s is already in use.", port)
+                except socket.gaierror as e:
+                    logger.error("restapi.listen might have a wrong value: '%s' -> %s", config["restapi"]["listen"], e)
+    for program in ["pg_ctl", "initdb", "pg_controldata", "pg_basebackup", "postgres"]:
+        if not find_executable(program, bin_dir):
+            logger.warning("Program '%s' not found.", program)
+    if "bootstrap" not in config:
+        logger.warning("bootstrap section is missing")
+    elif not isinstance(config["bootstrap"], dict):
+        logger.warning("bootstrap is not a dictionary")
+    else:
+        if "dcs" not in config["bootstrap"]:
+            logger.warning("bootstrap.dcs is not defined")
```

Comment:
This is absolutely fine, and moreover is the whole bootstrap is used only onces, when we bootstrap the new cluster.